datalog_compiler/src/lib.rs

```diff
@@ -0,0 +1,545 @@
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let outputs = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Output(_, ident) => Some(ident.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let rules = program
+        .rules
+        .iter()
+        .filter_map(|decl| match decl {
+            Declaration::Rule(rule) => Some(rule.clone()),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let mut flat_graph = FlatGraph::default();
+    let mut tee_counter = HashMap::new();
+    let mut merge_counter = HashMap::new();
+
+    let mut created_rules = HashSet::new();
+    for decl in program.rules {
+        let target_ident = match decl {
+            Declaration::Input(_, ident) => ident.clone(),
+            Declaration::Output(_, ident) => ident.clone(),
+            Declaration::Rule(rule) => rule.target.name.clone(),
+        };
+
+        if !created_rules.contains(&target_ident) {
+            created_rules.insert(target_ident.clone());
+            flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Named(
+                NamedHfStatement {
+                    name: syn::Ident::new(&target_ident.name, Span::call_site()),
+                    equals: Token![=](Span::call_site()),
+                    pipeline: Pipeline::Link(PipelineLink {
+                        lhs: Box::new(Pipeline::Operator(Operator {
+                            path: parse_quote!(merge),
+                            paren_token: Paren::default(),
+                            args: Punctuated::new(),
+                        })),
+                        connector: ArrowConnector {
+                            src: None,
+                            arrow: Token![->](Span::call_site()),
+                            dst: None,
+                        },
+                        rhs: Box::new(Pipeline::Operator(Operator {
+                            path: parse_quote!(tee),
+                            paren_token: Paren::default(),
+                            args: Punctuated::new(),
+                        })),
+                    }),
+                },
+            ));
+        }
+    }
+
+    for target in inputs {
+        let target_ident = syn::Ident::new(&target.name, Span::call_site());
+
+        let merge_index = merge_counter.entry(target.name.clone()).or_insert(0);
+        let my_merge_index = *merge_index;
+        *merge_index += 1;
+
+        flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
+            Pipeline::Link(PipelineLink {
+                lhs: Box::new(Pipeline::Operator(Operator {
+                    path: parse_quote!(recv_stream),
+                    paren_token: Paren::default(),
+                    args: vec![parse_quote!(#target_ident)]
+                        .iter()
+                        .cloned::<syn::Expr>()
+                        .collect(),
```

Comment:
cleaner: `Punctuated::from_iter([parse_quote!(#target_ident)])`

Also clone is unnecessary?